pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        processing.
+        """
+        if self.par is None:
+            return False
+        # TODO: At the moment, we can only perform any of the flat-field
+        # corrections if we are applying the pixel-flat correction.
+#        return self.par['use_pixelflat'] or self.par['use_specillum'] or self.par['use_illumflat']
+        return self.par['use_pixelflat']
 
-        Args:
-            force (bool, optional):
+    def apply_gain(self, force=False):
+        """
+        Use the gain to convert images from ADUs to electrons/counts.
 
-        Returns:
-            `numpy.ndarray`:  copy of self.image
+        Conversion applied to :attr:`image, :attr:`var`, and :attr:`rn2img`.
 
+        Args:
+            force (:obj:`bool`, optional):
+                Force the gain to be applied to the image, even if the step log
+                (:attr:`steps`) indicates that it already has been.
         """
         step = inspect.stack()[0][3]
-        # Check if already trimmed
-        if self.steps[step] and (not force):
-            msgs.warn("Gain was already applied. Returning")
-            return self.image.copy()
-
-        gain = np.atleast_1d(self.detector['gain']).tolist()
-        # Apply
-        self.image *= procimg.gain_frame(self.datasec_img, gain)
+        if self.steps[step] and not force:
+            # Already applied
+            msgs.warn('Gain was already applied.')
+            return
+        gain = procimg.gain_frame(self.datasec_img, np.atleast_1d(self.detector['gain']))
+        self.image *= gain
+        if self.var is not None:
+            # Convert the variance
+            self.var *= gain**2
+        if self.rn2img is not None:
+            # Convert the detector variance
+            self.rn2img *= gain**2
         self.steps[step] = True
-        # Return
-        return self.image.copy()
 
-    def build_ivar(self):
+#    def build_ivar(self):
+#        """
+#        Generate the inverse variance in the image.
+#
+#        This is a simple wrapper for :func:`~pypeit.core.procimg.variance_frame`.
+#
+#        Returns:
+#            `numpy.ndarray`_: The inverse variance in the image.
+#        """
+#        # Generate
+#        rawvarframe = procimg.variance_frame(self.datasec_img, self.image,
+#                                             self.detector['gain'], self.ronoise,
+#                                             # TODO: dark current is expected to
+#                                             # be in e-/s.
+#                                             darkcurr=self.detector['darkcurr'],
+#                                             # TODO: exptime is expected to be
+#                                             # in seconds
+#                                             exptime=self.exptime,
+#                                             rnoise=self.rn2img)
+#        return utils.inverse(rawvarframe)
+
+    def estimate_readnoise(self):
```

Comment:
I suggest we get rid of this. I don't see why would ever do it instead of just analysing bias frames which would be way more robust.

Reply:
As discussed over zoom, we decided to keep this an add a parameter that allows the user to select this.